### PR TITLE
feat: setup runtime types, setup game skeleton

### PR DESCRIPTION
**Pull request by weiijiie**
- Added runtime types
- Setup basic game skeleton based on [prototype](https://svelte.dev/repl/5ec93c82e4ab439b9e3d98862cc4de5a?version=3.43.0)
- Used [zustand](https://github.com/pmndrs/zustand) for state management -> very simple API based around mutating a global store
- Used [immer](https://github.com/immerjs/immer) to help make mutations simpler

Visit `/testing` and press `Load Example` to load the example JSON and display the game.

Currently ported the simple components from my prototype as `TempNarration`, `TempDialogue`, `TempImage`, and `TempClickable`. We can slowly replace them with the fully fleshed out components as we complete them.

### CODE REVIEW

File: schema/actions.ts [hunk 1]
```diff
@@ -12,9 +21,12 @@ import { Infer, number, optional, string, type } from 'superstruct'
...
-  type: string(),
-  duration: optional(number()),
-})
+export const ActionSchema = intersection([
+  type({
+    type: string(),
+    duration: defaulted(number(), 0),
+  }),
+  record(string(), any()),
+])
```
Comment (tlylt) [inline schema/actions.ts:30]: The comment for this seems outdated, as now the type could also be a string?

File: pages/testing.tsx [hunk 2]
```diff
@@ -26,34 +28,30 @@ const SchemaInput: React.FC = () => {
...
       {debouncedSchema !== '' && error && (
-        <p className="text-red-600 text-lg">Validation Error: {error}</p>
-      )}
-      {gameSchema && (
-        <pre className="p-2 max-w-2xl whitespace-pre-wrap border-2 border-green-500 shadow">
-          {JSON.stringify(gameSchema, null, 2)}
-        </pre>
+        <p className="text-lg text-red-600">Validation Error: {error}</p>
       )}
+      {game && <GameDisplay game={game}></GameDisplay>}
```
Comment (tlylt) [inline pages/testing.tsx:54]: perhaps can use self closing `<GameDisplay game={game} />`

File: stores/store.ts [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+        const { timing, timerIds } = accum
+        const { duration, args, sceneId, execute } = action
+
+        let finishAction: FinishAction = undefined
+        let newTimerIds = []
+
+        // schedule the action at the specified timing
+        let id = setTimeout(() => {
+          set((state) => {
+            let newGame = produce<Game>(state.game, (game) => {
```
Comment (tlylt) [inline stores/store.ts:72]: can this be const?

File: lib/scene.ts [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+  clickables: Clickable[]
+
+  intro: Action[]
+  outro: Action[]
+
+  state: State
+}
+
+export function makeScene(schema: SceneSchema): Scene {
+  let id = schema.id
```
Comment (tlylt) [inline lib/scene.ts:33]: can this be `const` ? or just replace the usage of `id` with `schema.id` below?

File: lib/actions.ts [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+
+        narration.shown = false
+      }
+    }
+  },
+
+  hideNarration: ({ args, game, sceneId }) => {
+    const { value } = args
+
+    let narration = getNarration(game, sceneId, value)
```
Comment (tlylt) [inline lib/actions.ts:85]: can possibly be `const` ?

**Review by tlylt — CHANGES_REQUESTED** [review]
Great job for setting up the overall game flow 👍 , I added some minor comments to address or clarify, and afterward it should be good to merge. We should be able to work off this implementation and make the individual components work well.

File: lib/actions.ts (hunk 5 above)
Comment (weiijiie) [reply lib/actions.ts:85]: yup, will change!

File: lib/scene.ts (hunk 4 above)
Comment (weiijiie) [reply lib/scene.ts:33]: yea, changing to const

File: stores/store.ts (hunk 3 above)
Comment (weiijiie) [reply stores/store.ts:72]: sure, changing to const too

File: pages/testing.tsx (hunk 2 above)
Comment (weiijiie) [reply pages/testing.tsx:54]: sure, updated!

File: schema/actions.ts (hunk 1 above)
Comment (weiijiie) [reply schema/actions.ts:30]: the comment still seems to be accurate to me? basically an action schema must have `type` and `duration` (or leave duration blank), and the rest of the fields can be captured as a record of string to any.

File: lib/actions.ts [hunk 6]
```diff
@@ -1,20 +1,28 @@
 import { ActionSchema } from '../schema/actions'
-import { Game, getClickable, getDialogue, getImage, getNarration } from './game'
+import { Clickable, Dialogue, Image, Narration } from './elements'
+import { Game } from './game'
+import { Scene } from './scene'
+
+// modify any property to have the changes persisted to the game state
+// if called via the `executeActions` method of the store
+export type ModifyableArgs = {
```
Comment (tlylt) [inline lib/actions.ts:8]: a minor nit, should be `ModifiableArgs` ?

Comment (weiijiie) [reply lib/actions.ts:8]: HAHA yeah

**Review by tlylt — APPROVED** [review]

**Pull request merged**